File: completable-reactor-graph-viewer/src/main/kotlin/ru/fix/completable/reactor/graph/viewer/gl/NodeCenter.java
package ru.fix.completable.reactor.graph.viewer.gl;

import javafx.beans.property.ReadOnlyDoubleWrapper;
import javafx.scene.Node;
import lombok.Getter;
import lombok.experimental.var;

import java.util.Optional;

/**
 * @author Kamil Asfandiyarov
 */
class NodeCenter {

    final Node world;
    final Node node;
    final Optional<CentrableNode> centrableNode;

    @Getter
    ReadOnlyDoubleWrapper centerX = new ReadOnlyDoubleWrapper();

    @Getter
    ReadOnlyDoubleWrapper centerY = new ReadOnlyDoubleWrapper();

    public NodeCenter(Node world, Node node) {
        this.world = world;
        this.node = node;

        if (node instanceof CentrableNode) {
            centrableNode = Optional.of((CentrableNode) node);
        } else {
            centrableNode = Optional.empty();
        }

        node.boundsInParentProperty().addListener((observableValue, oldBounds, newBounds) -> {

                    var pNode = node.getParent();
                    var diffx = 0.0;
                    var diffy = 0.0;

                    while (pNode != null && pNode != world) {
                        diffx += pNode.getLayoutX();
                        diffy += pNode.getLayoutY();
                        pNode = pNode.getParent();
                    }

                    double nodeCenterX;
                    double nodeCenterY;

                    if (centrableNode.isPresent()) {
                        nodeCenterX = centrableNode.get().getCenterX();
                        nodeCenterY = centrableNode.get().getCenterY();
                    } else {
                        nodeCenterX = newBounds.getMinX() + newBounds.getWidth() / 2;
                        nodeCenterY = newBounds.getMinY() + newBounds.getHeight() / 2;
                    }

                    centerX.set(nodeCenterX + diffx);
                    centerY.set(nodeCenterY + diffy);
                }
        );
    }
}

File: completable-reactor-graph-viewer/src/main/kotlin/ru/fix/completable/reactor/graph/viewer/gl/NodeCenter.kt
package ru.fix.completable.reactor.graph.viewer.gl;

import javafx.beans.property.ReadOnlyDoubleWrapper;
import javafx.scene.Node;

/**
 * @author Kamil Asfandiyarov
 */
class NodeCenter(world: Node, node: Node) {

    val centerX = ReadOnlyDoubleWrapper()
    val centerY = ReadOnlyDoubleWrapper()

    init {
        var centrableNode: CentrableNode?
        if (node is CentrableNode) {
            centrableNode = node
        } else {
            centrableNode = null
        }

        node.boundsInParentProperty().addListener { observableValue, oldBounds, newBounds ->

            var pNode = node.getParent();
            var diffx = 0.0
            var diffy = 0.0

            while (pNode != null && pNode != world) {
                diffx += pNode.layoutX;
                diffy += pNode.layoutY;
                pNode = pNode.parent;
            }

            val nodeCenterX: Double
            val nodeCenterY: Double

            if (centrableNode != null) {
                nodeCenterX = centrableNode.centerX
                nodeCenterY = centrableNode.centerY
            } else {
                nodeCenterX = newBounds.minX + newBounds.width / 2;
                nodeCenterY = newBounds.minY + newBounds.height / 2;
            }

            centerX.set(nodeCenterX + diffx)
            centerY.set(nodeCenterY + diffy)
        }
    }
}
